---
title: "Supereffectors_Analysis_PART2"
author: "Veronika Niederlova"
date: "3 8 2021"
output: html_document
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(include = TRUE, warning = FALSE, message = FALSE, error = TRUE, cache = TRUE)

library("Matrix")
library(Seurat)
library(DT)
library(dplyr)
library(here)
library(ggplot2)
library(kableExtra)
library(cowplot)
library(tidyverse)
library(reshape)
library(SingleR)
library(ProjecTILs)
library(ggprism)
library(patchwork)

convertHumanGeneList <- function(x){

require("biomaRt")
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")

genesV2 = getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = x , mart = human, attributesL = c("mgi_symbol"), martL = mouse, uniqueRows=T)

humanx <- unique(genesV2[, 2])

# Print the first 6 genes found to the screen
print(head(humanx))
return(humanx)
}

```

The input of this analysis is a processed Seurat file output of Supereffectors_Analysis_PART1. The file is called exp07_W3_Treg_processed_data_set_1_pre_filtering.rds. Please, download it from the GEO and save it to the `./data folder`. 

In this part of the analysis workflow , we will perform all necessary QC steps leading to a filtered final file, in particular:

- Annotation of cell types
- Filtering low quality and contaminating cells
- Normalization
- Removing the effect of cell cycle.

## Load the data


```{r}

path_to_downloaded_file <- "G://48_lab/Project Tregs DIA/! Manuscript in preparation/RNAseq DATA for upload/exp07_W3_Treg_processed_data_set_1_pre_filtering.rds"

E07_W3_510_data <- readRDS(path_to_downloaded_file)

```

The raw file contains  `r ncol(E07_W3_510_data)` cells and `r nrow(E07_W3_510_data)` genes.


## Annotation of cell types

We will use the SingleR package to annotate cell types. As a reference, we will use the Immgen microarray dataset and Mouse RNA seq dataset. Please note that to run this analysis, we use [the legacy version of the SingleR package](https://github.com/dviraran/SingleR).  

```{r}
# Annotations with SingleR

counts <- GetAssayData(E07_W3_510_data[["RNA"]])

E07_W3_510_data_singler <- CreateSinglerObject(counts=counts,
  project.name= "E07_Tregs", # choose
  min.genes = 200, # ignore cells with fewer than 200 transcripts
  technology = "10x", # choose
  species = "Mouse",
  citation = "Our 10x data exp 01-05", # choose
  normalize.gene.length = FALSE,        # needed for full-length platforms (e.g. smartseq)
  variable.genes = "de",  # see vignette
  fine.tune = FALSE, # TRUE would take very long
  reduce.file.size = TRUE, # leave out less-often used fields 
  do.signatures = FALSE,
  do.main.types = TRUE,
  numCores = SingleR.numCores)

## summary of main cell types

#ImmGen
summary(factor(E07_W3_510_data_singler$singler[[1]]$SingleR.single$labels))
summary(factor(E07_W3_510_data_singler$singler[[1]]$SingleR.single.main$labels))

#RNAseq
summary(factor(E07_W3_510_data_singler$singler[[2]]$SingleR.single$labels))
summary(factor(E07_W3_510_data_singler$singler[[2]]$SingleR.single.main$labels))


E07_W3_510_data <- AddMetaData(E07_W3_510_data,
E07_W3_510_data_singler$singler[[1]]$SingleR.single$labels,
                                 col.name = "Immgen_annot_single")

E07_W3_510_data <- AddMetaData(E07_W3_510_data,
E07_W3_510_data_singler$singler[[1]]$SingleR.single.main$labels,
                                 col.name = "Immgen_annot_single_main")

E07_W3_510_data <- AddMetaData(E07_W3_510_data,
E07_W3_510_data_singler$singler[[2]]$SingleR.single.main$labels,
                                 col.name = "MouseRNAseq_single_main")

E07_W3_510_data <- AddMetaData(E07_W3_510_data,
E07_W3_510_data_singler$singler[[2]]$SingleR.single$labels,
                                 col.name = "MouseRNAseq_single")
```

Based on the SingleR annotations, our sample contains the following cell types:

```{r fig.width=9, fig.height=3.5}

p1 <- DimPlot(E07_W3_510_data, group.by = "Immgen_annot_single_main") + ggtitle("Immgen main annotations") + theme(title = element_text(size = 11))
p2 <- DimPlot(E07_W3_510_data, group.by = "MouseRNAseq_single") + ggtitle("Mouse RNAseq annotations") + theme(title = element_text(size = 11))
(p1 + p2) + plot_annotation(
  title = 'Cell type annotations using the SingleR package',
  theme = theme(plot.title = element_text(hjust = 0.5)))


```

We see that clusters 8 and 11 contain cells annotated as non-T-cells. We will remove these clusters in the next steps.

## Filtering low quality and contaminating cells

We can also visualize the quality metrics in our cells:

```{r fig.width=8, fig.height=5}

p1 <- VlnPlot(E07_W3_510_data, features = c("nCount_RNA")) + ggtitle("RNA counts") + theme(title = element_text(size = 11)) + NoLegend()
p2 <- VlnPlot(E07_W3_510_data, features = c("nFeature_RNA")) + ggtitle("Gene counts") + theme(title = element_text(size = 11)) + NoLegend()
p3 <- VlnPlot(E07_W3_510_data, features = c("percent.mt")) + ggtitle("Percentage of mitochondrial genes") + theme(title = element_text(size = 11)) + NoLegend()
p4 <- VlnPlot(E07_W3_510_data, features = c("percent.rt")) + ggtitle("Percentage of ribosomal genes") + theme(title = element_text(size = 11)) + NoLegend()


(p1 + p2 + p3 + p4 + plot_layout(ncol = 2, byrow = T)) + plot_annotation(title = 'QC control of clusters',
  theme = theme(plot.title = element_text(hjust = 0.5)))

```

Based on these plots, we will filter our dataset for the following conditions:

- Remove cells with higher content of mitochondrial genes than 7.5%
- Remove clusters with non-T cell annotated cells (clusters 8 and 11)
- Remove clusters with mostly T-cell annotations, but low detected counts of UMIs/genes (clusters 8 and 9)


```{r}
E07_W3_510_data_filt <- subset(E07_W3_510_data, percent.mt < 7.5 & seurat_clusters %in% c(0:7,10))
```

The filteres Seurat object contains  `r ncol(E07_W3_510_data_filt)` cells and `r nrow(E07_W3_510_data_filt)` genes.

## Normalization, dimensional reduction

We will perform normalization, scaling, dimensional reduction and clustering using the filtered object. We use the following parameters:

- default Seurat parameters for NormalizeData and ScaleData
- selecting 1000 variable features
- PCA using 1000 variable features (12 principal components)
- umap based on 12 PCAs
- clustering with 0.3 resolution



```{r}
E07_W3_510_data_filt <- NormalizeData(E07_W3_510_data_filt, verbose = FALSE)
E07_W3_510_data_filt <- ScaleData(E07_W3_510_data_filt, verbose = FALSE)
E07_W3_510_data_filt <- FindVariableFeatures(E07_W3_510_data_filt, nfeatures = 1000, verbose = FALSE)
E07_W3_510_data_filt <- RunPCA(E07_W3_510_data_filt, npcs = 12, verbose = FALSE)
E07_W3_510_data_filt <- RunUMAP(E07_W3_510_data_filt, reduction = "pca", dims = 1:12)
E07_W3_510_data_filt <- FindNeighbors(E07_W3_510_data_filt, dims = 1:12)
E07_W3_510_data_filt <- FindClusters(E07_W3_510_data_filt, resolution = 0.3)

## Add new metadata column dered based on hashtags 
md <- E07_W3_510_data_filt@meta.data %>% mutate(dereg = case_when(hashtags %in% c("Mouse_Hashtag_5","Mouse_Hashtag_6","Mouse_Hashtag_7") ~ "DEREG-",
                                                                  TRUE ~ "DEREG+"))
E07_W3_510_data_filt@meta.data <- md
rownames(E07_W3_510_data_filt@meta.data) <- colnames(E07_W3_510_data_filt)
```

Here, we can see the resulting UMAP projection split by individual animals or experimental conditions (three mice DEREG+, three mice DEREG-).

```{r  fig.width=9, fig.height=3.5}

p1 <- DimPlot(E07_W3_510_data_filt, group.by = "hashtags") + ggtitle("UMAP by animals") + theme(title = element_text(size = 11))
p2 <- DimPlot(E07_W3_510_data_filt, group.by = "dereg") + ggtitle("UMAP by DEREG status") + theme(title = element_text(size = 11))

(p1 + p2 + plot_layout(ncol = 2, byrow = T)) + plot_annotation(title = 'UMAP - filtered sample',
  theme = theme(plot.title = element_text(hjust = 0.5)))


```



## Removing the effect of cell cycle.

Last, we will estimate the effect of cell cycle, as our dataset contained proliferating T cells as well as memory T cells.


```{r  fig.width=8, fig.height=3.5}
FeaturePlot(E07_W3_510_data_filt, features = c("Mki67","Pcna"))
```

We can see that most of the variability is driven by cell cycle. In the next step, we will use the Seurat estimation of cell cycle phase scores. These scores will help us regress out the effect of the cell cycle.

```{r fig.width=5, fig.height=3.5}
s.genes <- cc.genes.updated.2019$s.genes
 g2m.genes <- cc.genes.updated.2019$g2m.genes

 m.s.genes <- convertHumanGeneList(cc.genes.updated.2019$s.genes)
 m.g2m.genes <- convertHumanGeneList(cc.genes.updated.2019$g2m.genes)

seu_wocc <- CellCycleScoring(E07_W3_510_data_filt, s.features = m.s.genes, g2m.features = m.g2m.genes, set.ident = TRUE)
DimPlot(seu_wocc, group.by = "Phase") + ggtitle("Cell cycle phase")
```


```{r fig.width=10, fig.height=5}
seu_wocc <- ScaleData(seu_wocc, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(seu_wocc))


seu_wocc <- FindVariableFeatures(seu_wocc, selection.method = "vst", nfeatures = 600, verbose = FALSE)


seu_wocc <- RunPCA(seu_wocc, npcs = 12, verbose = FALSE)
seu_wocc <- RunUMAP(seu_wocc, reduction = "pca", dims = 1:12)
seu_wocc <- FindNeighbors(seu_wocc, dims = 1:12)
seu_wocc <- FindClusters(seu_wocc, resolution = 0.35)
  
p1 <- DimPlot(seu_wocc) + ggtitle("UMAP projection with clusters")
p2 <- DimPlot(seu_wocc, group.by = "dereg") + ggtitle("UMAP projection by DEREG status")

p1 + p2 

## In case you get mirror version of the paper UMAP:
# seu_wocc@reductions$umap@cell.embeddings <- seu_wocc@reductions$umap@cell.embeddings*-1
# DimPlot(seu_wocc)

```

This Seurat object was saved and used as the input for the next analysis (Supereffectors_PART3.Rmd).

# SessionInfo

```{r}
sessionInfo()
```

